### PR TITLE
Integrated barretenberg verification system into zk module with vkey upload support

### PR DESCRIPTION
**Pull request by Kushal7788**
Changes
- Added `proof_system` to VKey, MsgAddVKey, and MsgUpdateVKey (values: `groth16` default, `ultrahonk`).
- Added `ProofVerifyUltraHonk` query and `QueryVerifyUltraHonkRequest` (proof, public_inputs, vkey_name or vkey_id).
- Constants: ProofSystemGroth16, ProofSystemUltraHonk; ValidateBasic checks proof_system and validates vkey bytes for groth16.
- Genesis validation branches on proof_system (size-only for ultrahonk; full Circom validation for groth16).
- AddVKey / UpdateVKey / InitGenesis take and store proof_system; groth16 uses existing Circom JSON validation, ultrahonk uses barretenberg vkey validation.
- Implemented `ProofVerifyUltraHonk`: resolve vkey by name or ID, require ultrahonk type, split public_inputs into 32-byte chunks, call barretenberg verifier; return ProofVerifyResponse{Verified}.
- ProofVerifyUltraHonk tests: success (by name and ID), invalid request cases, Groth16 vkey rejected, wrong public inputs returns verified=false. Testdata loaded from x/zk/barretenberg/testdata/statics with skip when missing.
- No store migration: new field is backward compatible; existing VKeys read as groth16 when proof_system is empty.

### CODE REVIEW

File: x/zk/barretenberg/vkey.go [hunk 1]
```diff
@@ -94,26 +104,44 @@ func (vk *VerificationKey) IsClosed() bool {
...
 		return fmt.Errorf("%w: empty data", ErrInvalidVKey)
 	}
-
-	// Minimum reasonable size for an UltraHonk verification key
-	// This is a heuristic based on the expected structure
-	const minVKeySize = 1024
-	if len(data) < minVKeySize {
-		return fmt.Errorf("%w: data too small (%d bytes, minimum %d)", ErrInvalidVKey, len(data), minVKeySize)
+	if maxSizeBytes > 0 && uint64(len(data)) > maxSizeBytes {
+		return fmt.Errorf("%w: vkey size %d exceeds max %d", ErrInvalidVKey, len(data), maxSizeBytes)
```
Comment (Peartes) [inline x/zk/barretenberg/vkey.go:118]: I think we should use the errorsmod package from the sdk so we can bubble up the error code space

File: x/zk/barretenberg/vkey_test.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+		t.Skipf("test vector %s not found: %v", filename, err)
+	}
+	return data
+}
+
+// TestParseVerificationKeyEmpty tests parsing empty verification key.
+func TestParseVerificationKeyEmpty(t *testing.T) {
+	_, err := ParseVerificationKey(nil)
+	if err == nil {
+		t.Error("expected error for nil data")
```
Comment (Peartes) [inline x/zk/barretenberg/vkey_test.go:34]: this works for simple cases but it's probably better to test the error we're returning instead

File: x/zk/client/cli/tx.go [hunk 3]
```diff
@@ -55,17 +55,12 @@ $ %s tx zk add-vkey rollup_batch ./rollup_vkey.json "Rollup batch verification"
...
 
-			// Read vkey JSON file
+			// Read vkey file (Groth16 JSON or UltraHonk binary; validation in ValidateBasic)
 			vkeyBytes, err := os.ReadFile(args[1])
 			if err != nil {
 				return fmt.Errorf("failed to read vkey file: %w", err)
 			}
 
-			// Validate the vkey
-			if err := types.ValidateVKeyBytes(vkeyBytes, 0); err != nil {
```
Comment (Peartes) [inline x/zk/client/cli/tx.go:65]: why are we removing the validation ?

File: x/zk/barretenberg/vkey.go (hunk 1 above)
Comment (Peartes) [reply x/zk/barretenberg/vkey.go:118]: I see this is to be called from other functions so it's fine as it is

**Review by Peartes — APPROVED** [review]

**Review by crucible-burnt — COMMENTED** [review]
## 🔍 Crucible Security Review

### Summary
This PR adds UltraHonk (Barretenberg) proof system support alongside existing Groth16. The implementation follows established patterns and includes comprehensive validation at multiple layers.

### Security Assessment
- **Risk Level**: Medium (new cryptographic verification pathway)

**Findings:**

1. **Validation layering is sound**: Vkey validation properly cascades through `ValidateBasic()` → `ValidateVKeyForProofSystem()` → system-specific validators. Good defense-in-depth.

2. **`MinVKeySizeBytes` reduced from 1024 to 256** (`x/zk/barretenberg/vkey.go:11`): Verify this minimum is appropriate for the smallest legitimate UltraHonk vkey. If Barretenberg vkeys are typically much larger, consider a higher minimum to reject more garbage inputs before the native C call.

3. **Public inputs alignment check** (`x/zk/keeper/query_server.go:101`): Good — properly validates `len(publicInputs) % FieldElementSize != 0` before chunking. No out-of-bounds risk.

4. **Empty public inputs allowed**: `len(publicInputs) == 0` passes the modulo check (0 % 32 == 0), resulting in an empty `chunks` slice. Verify this is intentionally supported (circuits with no public inputs). Consider documenting this behavior.

5. **Vkey validation moved from CLI to ValidateBasic** (`x/zk/client/cli/tx.go`): This is a positive change — validation now happens at the msg level rather than only client-side. Ensures server-side validation regardless of client used.

6. **Gas metering maintained** for vkey operations in `AddVKey`/`UpdateVKey` — good.

7. **`proof_system` default to `groth16`** when empty: Correctly applied in keeper, genesis, and validation. Backward compatible.

8. **Test coverage removed then relocated** (`barretenberg/verifier_test.go` → `barretenberg/vkey_test.go`): Tests were moved, not deleted. Coverage maintained.

### Immunefi Pattern Check
- No singleton proto.Message patterns detected
- No JWT validation applicable
- Authority/signer validation properly maintained through existing keeper patterns
- No fee bypass vectors — gas metering is consistent

### False Report Risk
- The comment changes (e.g., `vkey_json` → `vkey_bytes`) correctly reflect the new dual-format support. No misleading comments detected.
- The `proof_system` field defaulting to empty string could confuse auditors — consider adding a comment that empty == groth16 for backward compat.

### Code Quality Notes
- Clean separation of Groth16 vs UltraHonk paths via `ValidateVKeyForProofSystem`
- Proper `defer vk.Close()` patterns for native resource cleanup
- `goerrors.Is(err, barretenberg.ErrVerificationFailed)` correctly distinguishes invalid proofs from system errors
- Tests are thorough including concurrent verification, round-trip, and edge cases

### Status
Well-structured addition. Medium risk due to new cryptographic pathway — recommend verifying the `MinVKeySizeBytes=256` threshold and documenting the empty public inputs behavior. No blocking issues found.

File: x/zk/client/cli/tx.go (hunk 3 above)
Comment (Kushal7788) [reply x/zk/client/cli/tx.go:65]: The above validations are inherently called when using `msg.ValidateBasic()` Validation was effectively run twice in the old flow because of an explicit CLI check
